### PR TITLE
Add cover grammar for 'await using'

### PR DESCRIPTION
**Pull request by rbuckton**
Adds a cover grammar for `await using`.

Fixes #12

### CODE REVIEW

**Comment by rbuckton** [comment]
This is a fairly early draft, but I'd like to get some feedback on the cover grammar. I'll continue to add some of the missing semantics rules needed to apply the supplemental syntax in later commits.

**Comment by rbuckton** [comment]
I still need to review `Contains` and the early errors for `delete` to determine if there is additional work to do to support the cover grammar. In the case of `Contains`, we do not currently explicitly handle `PrimaryExpression : CoverParenthesizedExpressionAndArrowParameterList`, though we do handle `ArrowParameters : CoverParenthesizedExpressionAndArrowParameterList`, so it is unclear whether something similar is needed.

**Comment by rbuckton** [comment]
Here are some potential test cases for this grammar:

```js
await using
```
is parsed as an _ExpressionStatement_ containing an _AwaitExpression_ (covered by _CoverAwaitExpressionAndAwaitUsingDeclarationHead_).

```js
await using
x = y
```
is parsed as an _ExpressionStatement_ containing an _AwaitExpression_ (covered by _CoverAwaitExpressionAndAwaitUsingDeclarationHead_), followed by an _ExpressionStatement_ containing an _AssignmentExpression_.

```js
await using x
```
is a syntax error. _CoverAwaitExpressionAndAwaitUsingDeclarationHead_ is parsed for `await using`, but an _IdentifierName_ cannot follow a UnaryExpression. This is then parsed as an _AwaitUsingDeclaration_, but an early error raises a syntax error due to the missing initializer.

```js
await
using x
```
is a syntax error. _CoverAwaitExpressionAndAwaitUsingDeclarationHead_ is parsed for `await using`, but an _IdentifierName_ cannot follow a UnaryExpression. When attempting to parse as an _AwaitUsingDeclaration_, this fails to meet the supplemental syntax for _AwaitUsingDeclarationHead_.

```js
await
using
x
```
is parsed as an _ExpressionStatement_ containing an _AwaitExpression_ (covered by _CoverAwaitExpressionAndAwaitUsingDeclarationHead_), followed by an _ExpressionStatement_ containing an _IdentifierReference_.

**Comment by rbuckton** [comment]
@waldemarhorwat: I'd appreciate if you're able to review sometime this week.

File: spec.emu [hunk 1]
```diff
@@ -2224,6 +2602,25 @@ contributors: Ron Buckton, Ecma International
...
+            It is a Syntax Error if the BoundNames of |BindingList| contains *"let"*.
+          </li>
+          <li>
+            It is a Syntax Error if the BoundNames of |BindingList| contains *"await"*.
+          </li>
+          <li>
+            It is a Syntax Error if the BoundNames of |BindingList| contains any duplicate entries.
+          </li>
+          <li>
+            It is a Syntax Error if the goal symbol is |Script| and |AwaitUsingDeclaration| is not contained, either directly or indirectly, within a |Block|, |CaseBlock|, |ForStatement|, |ForInOfStatement|, |FunctionBody|, |GeneratorBody|, |AsyncGeneratorBody|, |AsyncFunctionBody|, |ClassStaticBlockBody|, or |ClassBody|.
```
Comment (DanielRosenwasser) [inline spec.emu:2621]:
```suggestion
            It is a Syntax Error if the goal symbol is |Script| and |AwaitUsingDeclaration| is not contained, either directly or indirectly, within a |Block|, |CaseBlock|, |ForStatement|, |ForInOfStatement|, |FunctionBody|, |GeneratorBody|, |AsyncGeneratorBody|, |AsyncFunctionBody|, or |ClassStaticBlockBody|.
```

Comment (DanielRosenwasser) [reply spec.emu:2621]: Actually, let's keep that separate from this change.

File: spec.emu [hunk 2]
```diff
@@ -2179,36 +2542,51 @@ contributors: Ron Buckton, Ecma International
...
 
-        <ins>
+        <ins class="block">
         UsingDeclaration[In, Yield, Await] :
-          `using` [no LineTerminator here] [lookahead != `await`] BindingList[?In, ?Yield, ?Await, +Using] `;`
-          [+Await] `using` [no LineTerminator here] `await` [no LineTerminator here] BindingList[?In, ?Yield, +Await, +Using] `;`
+          `using` [no LineTerminator here] [lookahead != `await`] BindingList[?In, ?Yield, ?Await, ~Pattern] `;`
+
+        AwaitUsingDeclaration[In, Yield] :
+          CoverAwaitExpressionAndAwaitUsingDeclarationHead[?Yield] [no LineTerminator here] [lookahead != `await`] BindingList[?In, ?Yield, ?Await, ~Pattern] `;`
```
Comment (waldemarhorwat) [inline spec.emu]: Get rid of both [lookahead != `await`] restrictions here.

File: spec.emu [hunk 3]
```diff
@@ -2224,6 +2602,25 @@ contributors: Ron Buckton, Ecma International
...
+          AwaitUsingDeclaration :
+            CoverAwaitExpressionAndAwaitUsingDeclarationHead BindingList `;`
+        </emu-grammar>
+        <ul>
+          <li>|CoverAwaitExpressionAndAwaitUsingDeclarationHead| must cover an |AwaitUsingDeclarationHead|.</li>
+          <li>
+            It is a Syntax Error if the BoundNames of |BindingList| contains *"let"*.
+          </li>
+          <li>
+            It is a Syntax Error if the BoundNames of |BindingList| contains *"await"*.
```
Comment (waldemarhorwat) [inline spec.emu]: Get rid of the await restriction here.

File: spec.emu [hunk 4]
```diff
@@ -2474,24 +2868,24 @@ contributors: Ron Buckton, Ecma International
...
-          [+Await] `for` `await` `(` `var` ForBinding[?Yield, ?Await, <ins>~Using</ins>] `of` AssignmentExpression[+In, ?Yield, ?Await] `)` Statement[?Yield, ?Await, ?Return]
+          [+Await] `for` `await` `(` `var` ForBinding[?Yield, ?Await, <ins>+Pattern</ins>] `of` AssignmentExpression[+In, ?Yield, ?Await] `)` Statement[?Yield, ?Await, ?Return]
           [+Await] `for` `await` `(` <ins>[lookahead != `using` `of`]</ins> ForDeclaration[?Yield, ?Await, <ins>+Using</ins>] `of` AssignmentExpression[+In, ?Yield, ?Await] `)` Statement[?Yield, ?Await, ?Return]
 
         ForDeclaration[Yield, Await, <ins>Using</ins>] :
-          LetOrConst ForBinding[?Yield, ?Await, <ins>~Using</ins>]
-          <ins>[+Using] `using` [no LineTerminator here] [lookahead != `await`] ForBinding[?Yield, ?Await, +Using]</ins>
-          <ins>[+Using, +Await] `using` [no LineTerminator here] `await` [no LineTerminator here] ForBinding[?Yield, +Await, +Using]</ins>
+          LetOrConst ForBinding[?Yield, ?Await, <ins>+Pattern</ins>]
+          <ins>[+Using] `using` [no LineTerminator here] [lookahead != `await`] ForBinding[?Yield, ?Await, ~Pattern]</ins>
```
Comment (waldemarhorwat) [inline spec.emu]: Get rid of the [lookahead != `await`] restriction here.

File: spec.emu [hunk 5]
```diff
@@ -2962,6 +3356,37 @@ contributors: Ron Buckton, Ecma International
   <emu-clause id="sec-async-function-definitions">
     <h1>Async Function Definitions</h1>
 
+    <h2>Syntax</h2>
```
Comment (waldemarhorwat) [inline spec.emu:3356]: Why is a second definition of AsyncFunctionDeclaration pasted in the middle of the runtime semantics section? There already is a definition of AsyncFunctionDeclaration in the spec and I don't see the existing one being removed in this pull request.

Comment (rbuckton) [reply spec.emu:3356]: I'm not sure what you mean by this. There are no `<ins>` or `<del>` tags in the grammar below, so no changes were made. I included the grammar here for reference as `AwaitExpression` is referenced multiple times in the static semantics and runtime semantics.

**Review by waldemarhorwat — APPROVED** [review]

**Pull request merged**